### PR TITLE
Implement basic operations

### PR DESCRIPTION
**Pull request by achour94**

### CODE REVIEW

**Review by TheMaskedTurtle — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+            <scope>runtime</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-actuator</artifactId>
+            <scope>runtime</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.micrometer</groupId>
+            <artifactId>micrometer-registry-prometheus</artifactId>
```
Comment (FranckLecuyer) [inline pom.xml:136]: Not used

File: src/test/resources/logback-test.xml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            </pattern>
+        </encoder>
+    </appender>
+    <root level="info">
+        <appender-ref ref="STDOUT"/>
+    </root>
+    <!--
+        our logs
+    -->
+    <logger name="org.gridsuite.useradmin.server" level="trace"/>
```
Comment (FranckLecuyer) [inline src/test/resources/logback-test.xml:40]: org.gridsuite.spreadsheetconfig.server

File: src/main/java/org/gridsuite/spreadsheetconfig/server/entities/CustomColumnEntity.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+/**
+ * @author Achour BERRAHMA <achour.berrahma at rte-france.com>
+ */
+@Entity
+@Table(name = "spreadsheet_custom_column")
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
```
[the remark this answers is not in the document]
Comment (achour94) [reply src/main/java/org/gridsuite/spreadsheetconfig/server/entities/CustomColumnEntity.java:23]: We use the builder in the code (mapper and service)

File: src/main/java/org/gridsuite/spreadsheetconfig/server/entities/SpreadsheetConfigEntity.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
+/**
+ * @author Achour BERRAHMA <achour.berrahma at rte-france.com>
+ */
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
```
[the remark this answers is not in the document]
Comment (achour94) [reply src/main/java/org/gridsuite/spreadsheetconfig/server/entities/SpreadsheetConfigEntity.java:22]: We use the builder in the code (mapper and service)

File: src/main/java/org/gridsuite/spreadsheetconfig/server/entities/CustomColumnEntity.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    @Column(name = "id")
+    private UUID id;
+
+    @Column(name = "name", nullable = false, columnDefinition = "varchar(255)")
+    private String name;
+
+    @Column(name = "formula", columnDefinition = "varchar(255)")
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/spreadsheetconfig/server/entities/CustomColumnEntity.java:34]: Use CLOB for the formula column type

**Review by FranckLecuyer — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/spreadsheetconfig/server/entities/SpreadsheetConfigEntity.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    @Column(name = "id")
+    private UUID id;
+
+    @Column(name = "sheet_type", nullable = false)
+    @Enumerated(EnumType.STRING)
+    private SheetType sheetType;
+
+    @OneToMany(mappedBy = "spreadsheetConfig", cascade = CascadeType.ALL, orphanRemoval = true)
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/spreadsheetconfig/server/entities/SpreadsheetConfigEntity.java:38]: Why not using here the annotations ElementCollection and CollectionTable
and the Embeddable annotation in CustomColumnEntity (and removing the ManyToOne relationship with SpreadsheetConfigEntity in CustomColumnEntity)

**Review by FranckLecuyer — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/spreadsheetconfig/server/controller/SpreadsheetConfigController.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+public class SpreadsheetConfigController {
+
+    private final SpreadsheetConfigService spreadsheetConfigService;
+
+    public static final String DUPLICATE_FROM = "duplicateFrom";
+
+    @PostMapping(consumes = APPLICATION_JSON_VALUE)
+    @Operation(summary = "Create a new spreadsheet configuration",
+            description = "Creates a new spreadsheet configuration and returns its ID")
+    @ApiResponse(responseCode = "201", description = "Configuration created",
```
Comment (AbdelHedhili) [inline src/main/java/org/gridsuite/spreadsheetconfig/server/controller/SpreadsheetConfigController.java:45]: Why do we start using more specific return code ? We used to stick to 200 for OK (even for post). For the sake of harmonization shouldn't we keep the same logic everywhere (or change it everywhere but not here in this PR) ?

File: src/main/java/org/gridsuite/spreadsheetconfig/server/dto/SpreadsheetConfigDto.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+import java.util.UUID;
+
+/**
+ * @author Achour BERRAHMA <achour.berrahma at rte-france.com>
+ */
+@Value
+@Builder
+@Jacksonized
+@Schema(name = "SpreadsheetConfigDto", description = "Spreadsheet configuration")
+public class SpreadsheetConfigDto {
```
Comment (AbdelHedhili) [inline src/main/java/org/gridsuite/spreadsheetconfig/server/dto/SpreadsheetConfigDto.java:26]: same
```suggestion
public class SpreadsheetConfigInfos {
```

**Review by AbdelHedhili — CHANGES_REQUESTED** [review]

File: .mvn/lombok-config-copy.marker
[the remark this answers is not in the document]
Comment (achour94) [reply .mvn/lombok-config-copy.marker:1]: It's used after, so sonar will ignore the lombok annotation in coverage analysis

File: src/main/java/org/gridsuite/spreadsheetconfig/server/controller/SpreadsheetConfigController.java (hunk 7 above)
Comment (achour94) [reply src/main/java/org/gridsuite/spreadsheetconfig/server/controller/SpreadsheetConfigController.java:46]: We can keep it like this here, it's a good practice to return the right code :)

**Review by AbdelHedhili — APPROVED** [review]

**Review by FranckLecuyer — APPROVED** [review]

File: src/test/java/org/gridsuite/spreadsheetconfig/server/SpreadsheetConfigControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+
+    @Test
+    void testDuplicate() throws Exception {
+        SpreadsheetConfigInfos configToCreate = new SpreadsheetConfigInfos(null, SheetType.BATTERIES, createCustomColumns());
+        UUID configUuid = postSpreadsheetConfig(configToCreate);
+
+        mockMvc.perform(post(URI_SPREADSHEET_CONFIG_BASE + "/duplicate")
+                        .queryParam("duplicateFrom", UUID.randomUUID().toString()))
+                .andExpect(status().isNotFound());
+
```
Comment (TheMaskedTurtle) [inline src/test/java/org/gridsuite/spreadsheetconfig/server/SpreadsheetConfigControllerTest.java:213]: This is a separate test testDuplicateNonExistent

File: src/test/java/org/gridsuite/spreadsheetconfig/server/SpreadsheetConfigControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+        UUID configUuid = saveAndReturnId(configToUpdate);
+
+        SpreadsheetConfigInfos updatedConfig = new SpreadsheetConfigInfos(configUuid, SheetType.BUSES, createUpdatedCustomColumns());
+
+        String updatedConfigJson = mapper.writeValueAsString(updatedConfig);
+
+        mockMvc.perform(put(URI_SPREADSHEET_CONFIG_GET_PUT + configUuid)
+                        .content(updatedConfigJson)
+                        .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isNoContent());
```
Comment (TheMaskedTurtle) [inline src/test/java/org/gridsuite/spreadsheetconfig/server/SpreadsheetConfigControllerTest.java:160]: This is a separate test

File: src/test/java/org/gridsuite/spreadsheetconfig/server/SpreadsheetConfigControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+        UUID configUuid = saveAndReturnId(configToRead);
+
+        MvcResult receivedMetadata = mockMvc.perform(get(URI_SPREADSHEET_CONFIG_BASE + "/metadata")
+                        .queryParam("ids", configUuid.toString()))
+                .andExpect(status().isOk())
+                .andReturn();
+
+        List<MetadataInfos> metadata = mapper.readValue(
+                receivedMetadata.getResponse().getContentAsString(),
+                new TypeReference<List<MetadataInfos>>() { });
```
Comment (TheMaskedTurtle) [inline src/test/java/org/gridsuite/spreadsheetconfig/server/SpreadsheetConfigControllerTest.java:115]: You can put that in a private method ?

File: src/test/java/org/gridsuite/spreadsheetconfig/server/SpreadsheetConfigControllerTest.java [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+/**
+ * @author Achour BERRAHMA <achour.berrahma at rte-france.com>
+ */
+@SpringBootTest
+@AutoConfigureMockMvc
+class SpreadsheetConfigControllerTest {
```
Comment (TheMaskedTurtle) [inline src/test/java/org/gridsuite/spreadsheetconfig/server/SpreadsheetConfigControllerTest.java:39]: These are integration tests and not unit tests of the controller, so please rename the file SpeadsheetConfigIntegrationTest

**Review by TheMaskedTurtle — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/spreadsheetconfig/server/dto/CustomColumnInfos.java [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+@Schema(name = "CustomColumnDto", description = "Custom column configuration")
+public record CustomColumnInfos(
+
+    @NotNull(message = "Column name is mandatory")
+    @Schema(description = "Column name")
+    String name,
+
+    @Schema(description = "Column formula")
+    String formula
+) { }
```
Comment (Tristan-WorkGH) [inline src/main/java/org/gridsuite/spreadsheetconfig/server/dto/CustomColumnInfos.java:24]: It miss the type of data (string, interger, boolean, ...)
```suggestion

    @Schema(description = "Column type")
    ColumnType type;
) { }
```

Comment (achour94) [reply src/main/java/org/gridsuite/spreadsheetconfig/server/dto/CustomColumnInfos.java:24]: How can we get this information !?

File: src/test/java/org/gridsuite/spreadsheetconfig/server/SpreadsheetConfigControllerTest.java (hunk 10 above)
Comment (achour94) [reply src/test/java/org/gridsuite/spreadsheetconfig/server/SpreadsheetConfigControllerTest.java:160]: Not really

**Review by TheMaskedTurtle — APPROVED** [review]

**Pull request merged**